src/pint/models/binary_piecewise.py

```diff
@@ -0,0 +1,325 @@
+            )
+        for index in indices:
+            index_rf = f"{int(index):04d}"
+            for prefix in ["T0X_", "A1X_", "PLB_", "PUB_"]:
+                self.remove_param(prefix + index_rf)
+        self.setup()
+        self.validate()
+
+    def add_piecewise_param(self, param, param_unit, paramx, j):
+        if int(j) in self.get_prefix_mapping_component("X_"):
+            raise ValueError(
+                "Index '%s' is already in use in this model. Please choose another." % j
+            )
+        if j is None:
+            dct = self.get_prefix_mapping_component(param + "X_")
+            j = np.max(list(dct.keys())) + 1
+        i = f"{int(j):04d}"
+        if param is "A1":
+            self.add_param(
+                prefixParameter(
+                    name=param + "X_{0}".format(i),
+                    units=param_unit,
+                    value=paramx,
+                    unit_template=lambda x: param_unit,
+                    description="Parameter" + param + "variation",
+                    description_template=lambda x: param,
+                    parameter_type="float",
+                    frozen=False,
+                )
+            )
+        elif param is "T0":
+            self.add_param(
+                prefixParameter(
+                    name=param + "X_{0}".format(i),
+                    units=param_unit,
+                    value=paramx,
+                    unit_template=lambda x: param_unit,
+                    description="Parameter" + param + "variation",
+                    description_template=lambda x: param,
+                    parameter_type="float",
+                    frozen=False,
+                )
+            )
+
+        # self.setup()
+
+    def setup(self):
+        # print(self.PB)
+        super(BinaryBTPiecewise, self).setup()
+        # print(self.PB)
+        for bpar in self.params:
+            self.register_deriv_funcs(self.d_binary_delay_d_xxxx, bpar)
+        # Setup the model isinstance
+        self.binary_instance = self.binary_model_class()
+        # Setup the FBX orbits if FB is set.
+        # TODO this should use a smarter way to set up orbit.
+        T0X_mapping = self.get_prefix_mapping_component("T0X_")
+        # print(T0X_mapping)
+        T0Xs = {}
+        A1X_mapping = self.get_prefix_mapping_component("A1X_")
+        A1Xs = {}
+        XR1_mapping = self.get_prefix_mapping_component("PLB_")
+        XR1s = {}
+        XR2_mapping = self.get_prefix_mapping_component("PUB_")
+        XR2s = {}
+        # print(f'XR2:{XR2_mapping.values()}')
+        for t0n in T0X_mapping.values():
+            # print("hello from T0 mapping")
+            T0Xs[t0n] = getattr(self, t0n).quantity
+        # if None in T0Xs.values():
+        # print("Group with non-defined T0X, applying default T0 to group")
+        # TODO set default T0 value
+        if None not in T0Xs.values():
```

Comment:
Doesn't this skip defining all the parameters if any of the values is unset? That doesn't seem right?